j2cl-tasks/src/main/java/com/vertispan/j2cl/build/provided/TranslationsFileProcessor.java

```diff
@@ -97,17 +98,109 @@ public Optional<File> getTranslationsFile(List<Input> inputs, BuildLog log) {
                 throw new RuntimeException(e);
             }
-            log.warn("No matching locales for " + locale);
+            context.log().warn("No matching locales for " + locale);
             return Optional.empty();
         }
 
+
+        private Set<String> locales(String locale) {
+            Set<String> result = new HashSet<>();
+            result.add(locale);
+
+            if (locale.contains("_")) {
+                // according to spec, locale discaration can have 3 parts max
+                if(locale.indexOf("_") != locale.lastIndexOf("_")) {
+                    result.add(locale.substring(0, locale.lastIndexOf("_")));
+                }
+                result.add(locale.substring(0, locale.indexOf("_")));
+            }
+            return result;
+        }
+
+        private Optional<File> mergeFiles(HashMap<String, Set<NodeList>> suitableFiles, Set<String> locales, TaskContext context) {
+            Map<String, Node> resultedCodeSet = new HashMap<>();
+            // parent nodes first
+            locales.stream().sorted((o1, o2) -> {
+                if(o1.split("_").length > o2.split("_").length) {
+                    return 1;
+                }
+                return -1;
+            }).forEach(locale -> {
+                suitableFiles.get(locale).forEach(node -> {
+                    NodeList children = node.item(0).getChildNodes();
+
+                    for (int i = 0; i < children.getLength(); i++) {
+                        if (children.item(i).getNodeName().equals("translation")) {
+                            String key = children.item(i).getAttributes().getNamedItem("id").getNodeValue();
+                            // override parent keys if already exists
+                            resultedCodeSet.put(key, children.item(i));
+                        }
+                    }
+                });
+            });
+
+            Path folder = context.outputPath();
+            File generated = folder.resolve("generated_messages.xtb").toFile();
+            generateAndWriteXTB(resultedCodeSet, generated);
+            return Optional.of(generated);
+        }
+
+        private void generateAndWriteXTB(Map<String, Node> resultedCodeSet, File generated) {
+            StringBuffer sb = new StringBuffer();
+
+            sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
+            sb.append(System.lineSeparator());
+            sb.append("<!DOCTYPE translationbundle SYSTEM \"translationbundle.dtd\">");
+            sb.append(System.lineSeparator());
+
+            sb.append("<translationbundle lang=\"" + locale + "\">");
+            sb.append(System.lineSeparator());
+
+            resultedCodeSet.values().forEach(node -> {
+                sb.append("  <translation id=\"");
+                sb.append(node.getAttributes().getNamedItem("id").getNodeValue());
+                sb.append("\" key=\"");
+                sb.append(node.getAttributes().getNamedItem("key").getNodeValue());
```

Comment:
these key values need to be escaped right? a key could contain something not legal in an xml attr?

Reply:
This is still wrong, build fails if you make a `key` equal something like `key="MSG_&quot;test16"`, it is generated into the merged xtb file as `key="MSG_"test16"` instead of re-escaping it. Your `escape()` method below should be enough to deal with this?